staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json_test.go

```diff
@@ -1043,6 +1046,271 @@ func TestEncode(t *testing.T) {
+								"name":      "pod2",
+								"namespace": "default",
+							},
+						},
+					},
+				},
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			normalBuffer.Reset()
+			streamingBuffer.Reset()
+			if err := streamingSerializer.Encode(tc.in, &streamingBuffer); err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			if err := normalSerializer.Encode(tc.in, &normalBuffer); err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			matchEncodedJSON(t, streamingBuffer.Bytes(), normalBuffer.Bytes(), tc.exactMatch)
+		})
+	}
+}
+
+func TestFuzzStreamingCollectionsEncoding(t *testing.T) {
+	disableFuzzFieldsV1 := func(field *metav1.FieldsV1, c fuzz.Continue) {}
+	fuzzUnstructuredList := func(list *unstructured.UnstructuredList, c fuzz.Continue) {
+		list.Object = map[string]interface{}{
+			"kind":         "List",
+			"apiVersion":   "v1",
+			c.RandString(): c.RandString(),
+			c.RandString(): c.RandUint64(),
+			c.RandString(): c.RandBool(),
+			"metadata": map[string]interface{}{
+				"resourceVersion":    fmt.Sprintf("%d", c.RandUint64()),
+				"continue":           c.RandString(),
+				"remainingItemCount": fmt.Sprintf("%d", c.RandUint64()),
+				c.RandString():       c.RandString(),
+			}}
+		c.Fuzz(&list.Items)
+	}
+	fuzzMap := func(kvs map[string]interface{}, c fuzz.Continue) {
+		kvs[c.RandString()] = c.RandBool()
+		kvs[c.RandString()] = c.RandUint64()
+		kvs[c.RandString()] = c.RandString()
+	}
+	f := fuzz.New().Funcs(disableFuzzFieldsV1, fuzzUnstructuredList, fuzzMap)
+	streamingSerializer := json.NewSerializerWithOptions(json.DefaultMetaFactory, nil, nil, json.SerializerOptions{StreamingCollectionsEncoding: true})
+	streamingBuffer := &bytes.Buffer{}
+	normalSerializer := json.NewSerializerWithOptions(json.DefaultMetaFactory, nil, nil, json.SerializerOptions{StreamingCollectionsEncoding: false})
+	normalBuffer := &bytes.Buffer{}
+	t.Run("CarpList", func(t *testing.T) {
+		for i := 0; i < 1000; i++ {
+			list := &testapigroupv1.CarpList{}
+			f.Fuzz(list)
+			streamingBuffer.Reset()
+			normalBuffer.Reset()
+			if err := streamingSerializer.Encode(list, streamingBuffer); err != nil {
+				t.Fatal(err)
+			}
+			if err := normalSerializer.Encode(list, normalBuffer); err != nil {
+				t.Fatal(err)
+			}
+			matchEncodedJSON(t, streamingBuffer.Bytes(), normalBuffer.Bytes(), true)
+		}
+	})
+	t.Run("UnstructuredList", func(t *testing.T) {
+		for i := 0; i < 1000; i++ {
+			list := &unstructured.UnstructuredList{}
+			f.Fuzz(list)
+			streamingBuffer.Reset()
+			normalBuffer.Reset()
+			if err := streamingSerializer.Encode(list, streamingBuffer); err != nil {
+				t.Fatal(err)
+			}
+			if err := normalSerializer.Encode(list, normalBuffer); err != nil {
+				t.Fatal(err)
+			}
+			matchEncodedJSON(t, streamingBuffer.Bytes(), normalBuffer.Bytes(), false)
+		}
+	})
+}
+
+func matchEncodedJSON(t *testing.T, first, second []byte, exactMatch bool) {
+	t.Helper()
+	if exactMatch {
```

Comment:
I think we should start by aiming for exact match for everything, not accepting byte-differences if possible

Reply:
Problem is that unstructuredList has undeterministic field order due to usage of map. Not sure how we could test it without modifying the code to be deterministic.

Reply:
I'd expect to make the output deterministic like stdlib encoding... accumulate the keys, sort them, then iterate over them to output?

Reply:
Done